src/rule_definition_tools.jl

```diff
@@ -79,69 +161,84 @@ macro scalar_rule(call, maybe_setup, partials...)
             call.args[i] = esc(arg)
         end
     end
-    if all(Meta.isexpr(partial, :tuple) for partial in partials)
-        input_rep = :(first(promote($(inputs...))))  # stand-in with the right type for an input
-        forward_rules = Any[rule_from_partials(input_rep, partial.args...) for partial in partials]
-        reverse_rules = Any[]
-        for i in 1:length(inputs)
-            reverse_partials = [partial.args[i] for partial in partials]
-            push!(reverse_rules, rule_from_partials(inputs[i], reverse_partials...))
+
+    # For consistency in code that follows we make all partials tuple expressions
+    partials = map(partials) do partial
+        if Meta.isexpr(partial, :tuple)
+            partial
+        else
+            @assert length(inputs) == 1
+            Expr(:tuple, partial)
+        end
+    end
+
+    ############################################################################
+    # Main body: defining the results of the frule/rrule
+
+    # An expression that when evaluated will return the type of the input domain.
+    # Multiple repetitions of this expression should optimize ot. But if it does not then
+    # may need to move its definition into the body of the `rrule`/`frule`
+    𝒟 = :(typeof(first(promote($(call.args[2:end]...)))))
+
+    n_outputs = length(partials)
+    n_inputs = length(inputs)
```

Comment:
Since we are computing `n_inputs` here anyway, can we compute it earlier and use it above when making `partials` a tuple? Can't see exactly how we'd re-write it -- but just a thought.

Reply:
Do you mean for theck of `length(inputs)==1` ?
yeah maybe

idea of `n_inputs` (and `n_outputs`) is to be clear on shape of what is going on with in the pullback and pushforward definitions and what lines up with what.